File: point_controller/test/test_control.cpp
#include "point_controller/control.hpp"

#include <gtest/gtest.h>

#include <cmath>

using point_controller::ControllerConfig;
using point_controller::GoalPose;
using point_controller::RobotPose;
using point_controller::compute_command;
using point_controller::normalize_angle;

TEST(PointControllerControl, NormalizeAngleWrapsToPiRange)
{
  EXPECT_LT(normalize_angle(3.5), M_PI);
  EXPECT_GT(normalize_angle(-3.5), -M_PI);
}

TEST(PointControllerControl, ControllerAlignsBeforeDriving)
{
  const auto result = compute_command(
    RobotPose{0.0, 0.0, M_PI / 2.0},
    GoalPose{1.0, 0.0, 0.0},
    ControllerConfig{});

  EXPECT_EQ(result.phase, "align_heading");
  EXPECT_DOUBLE_EQ(result.v, 0.0);
  EXPECT_LT(result.w, 0.0);
}

TEST(PointControllerControl, ControllerDrivesWhenHeadingIsClose)
{
  const auto result = compute_command(
    RobotPose{0.0, 0.0, 0.01},
    GoalPose{1.0, 0.0, 0.0},
    ControllerConfig{});

  EXPECT_EQ(result.phase, "drive");
  EXPECT_GT(result.v, 0.0);
}

TEST(PointControllerControl, ControllerReachesInsideTolerances)
{
  const auto result = compute_command(
    RobotPose{0.0, 0.0, 0.0},
    GoalPose{0.01, 0.0, 0.01},
    ControllerConfig{});

  EXPECT_TRUE(result.reached);
  EXPECT_DOUBLE_EQ(result.v, 0.0);
  EXPECT_DOUBLE_EQ(result.w, 0.0);
}

File: point_controller/test/test_control.py
from math import pi

from point_controller.control import (
    ControllerConfig,
    GoalPose,
    RobotPose,
    compute_command,
    normalize_angle,
)


def test_normalize_angle_wraps_to_pi_range():
    assert normalize_angle(3.5) < pi
    assert normalize_angle(-3.5) > -pi


def test_controller_aligns_before_driving():
    result = compute_command(
        RobotPose(0.0, 0.0, pi / 2.0),
        GoalPose(1.0, 0.0, 0.0),
        ControllerConfig(),
    )

    assert result.phase == 'align_heading'
    assert result.v == 0.0
    assert result.w < 0.0


def test_controller_drives_when_heading_is_close():
    result = compute_command(
        RobotPose(0.0, 0.0, 0.01),
        GoalPose(1.0, 0.0, 0.0),
        ControllerConfig(),
    )

    assert result.phase == 'drive'
    assert result.v > 0.0


def test_controller_reaches_inside_tolerances():
    result = compute_command(
        RobotPose(0.0, 0.0, 0.0),
        GoalPose(0.01, 0.0, 0.01),
        ControllerConfig(),
    )

    assert result.reached
    assert result.v == 0.0
    assert result.w == 0.0
